man/sslclient.1
.TH sslclient 1
.SH NAME
sslclient \- setup a TLS client connection
.SH SYNOPSIS
.B sslclient 
[
.I opts
]
.I host 
.I port 
.I prog 
.SH DESCRIPTION
.I opts 
is a series of getopt-style options,
.I host 
is a host name for the client to connect to, and
.I prog 
is one or more arguments specifying a program to run 
for each successful connection. 

.B sslclient 
attempts to connect to a TCP server at 
.I host
.IR port . 
If the connection succeeds, 
.B sslclient 
runs 
.IR prog , 
with file descriptors 6 and 7 reading from and 
writing to a child process ssl. 
The ssl process attempts an SSL connect via the network. 
If it succeeds, it translates data between 
.I prog 
and the network, performing any necessary SSL encoding and decoding. 

Before running 
.I prog, 
.B sslclient 
sets certain environment variables. 

.SH OPTIONS
.B General Options: 
.TP
.B \-q
Quiet. Do not print error messages. 
.TP
.B \-Q
(Default.) Print error messages. 
.TP
.B \-v 
Verbose. Print error messages and status messages.
.P
.B Connection options: 
.TP
.B \-T \fIx+y
Give up on the connection attempt or SSL connection attempt after 
.I x+y 
seconds. The default value is: 2+58. When a 
.I host 
has several IP addresses,
.B sslclient 
tries to connect to the first IP address, waits 
.I x 
seconds, tries to connect to the second IP address, waits 
- Ix 
seconds, etc.; then it retries each address that timed out, waiting 
.I y 
seconds per address. You may omit 
.I +y 
to skip the second try. 
.TP
.B \-i \fIlocalip
Use 
.I localip 
as the IP address for the local side of the connection; quit if 
.I localip 
is not available. Normally 
.B sslclient 
lets the operating system choose an address. 
.TP
.B \-p \fIlocalport
Use 
.I localport
as the TCP port for the local side of the connection; quit if 
.I localport 
is not available. Normally 
.B sslclient 
lets the operating system choose a port . 
.TP
.B \-d
Delay sending data for a fraction of a second whenever 
the remote host is responding slowly. 
This is currently the default, but it may not be in the future; 
if you want it, set it explicitly. 
.TP 
.B \-D
Never delay sending data; enable TCP_NODELAY. 

.P
.B X509 certificate handling:
.TP
.B \-3
Read a null-terminated 
.I key password 
from file descriptor 3. 
.TP
.B \-a \fIcafile
Override the compiled-in CA file name. 
The CA file contains the list of CAs used to verify the server certificate. 
.TP
.B \-A \fIcadir
Override the compiled-in CA directory name. 
The CA directory contains certificates files used 
to verify the client certificate. This list augments the list from 
.I \-a \fIcafile. 
.TP
.B \-c \fIcertfile
Use the client certificate in 
.IR certfile . 
.TP
.B \-k \fIkeyfile
Use the client certificate key in 
.IR keyfile . 
.TP
.B \-V \fIverifydepth
Verify the server certificate chain to depth 
.IR verifydepth . 
The default value is 1. 
.TP
.B \-z \fIciphers 
Use the cipher list specified in 
.IR ciphers . 
.TP
.B \-x
(Default.) Verify the server certificate. 
.TP
.B \-X 
Do not verify the server certificate.
.TP
.B \-n 
(Default.) Verify that the server host name matches 
the FQDN provided in the certificate.
.TP
.B \-N
Do not verify that the server host name matches 
the FQDN provided in the certificate.

.P
.B Data-gathering options: 
.TP
.B \-h
(Default.) Look up the remote host name in DNS 
to set the environment variable $SSLREMOTEHOST. 
.TP
.B \-H
Do not look up the remote host name in DNS; 
remove the environment variable $SSLREMOTEHOST. 
.TP
.B \-l \fIlocalname
Do not look up the local host name in DNS; use 
.I localname 
for the environment variable $SSLLOCALHOST. 
A common choice for
.I localname 
is 0. 
.TP
.B \-r 
Attempt to obtain $SSLREMOTEINFO from the remote host. 
.TP
.B \-R 
(Default.) Do not attempt to obtain $SSLREMOTEINFO from the remote host. 
.TP
.B \-t \fIn
Give up on the $SSLREMOTEINFO connection attempt after 
.I n 
seconds. The default value is: 26. 
.TP
.B \-w \fIn
Give up on a connection or program after waiting 
.I n 
seconds for read or write. The default value is: 3600. 
.TP
.B \-4
Fall back to IPv4 sockets.  This is necessary for terminally broken
systems like OpenBSD which will not let IPv6 sockets connect to
V4-mapped IPv6 addresses.  Please note that this also applies to DNS
lookups, so you will have to use an DNS resolver with an IPv6 address to
connect to IPv6 systems.  Use \fBDNSCACHEIP\fR to set the DNS resolver
IP dynamically.
.TP
.B \-6
Force IPv6 mode in UCSPI environment variables, even for
IPv4 connections.  This will set \fB$PROTO\fR to \fBSSL6\fR and put
IPv4-mapped IPv6 addresses in \fBTCPLOCALIP\fR and \fBTCPREMOTEIP\fR.
.TP
.B \-s 
Store client and server certificate information in the environment, a la mod_ssl.
.TP 
.B \-S
(Default.) Do not store client and server certificate information in the environment. 
.TP
.B \-e 
Set protocol environment a la 
.BR tcpserver . 
Set $TCPLOCALIP, $TCPLOCALPORT, $TCPLOCALHOST, $TCPREMOTEIP, 
$TCPREMOTEPORT, $TCPREMOTEHOST, and $TCPREMOTEINFO 
from the corresponding $SSL variables. 
.TP
.B \-E 
(Default.) Do not set any 
.B tcpserver 
environment variables.

.SH SEE ALSO
sslserver(1),
sslconnect(1),
sslcat(1),
https@(1),
ucspi-tls(2),
tcp-environ(5).

http://www.superscript.com/ucspi-ssl/sslclient.html
http://httpd.apache.org/docs/2.0/mod/mod_ssl.html





doc/CHANGES
20011209
	Initial release.
	version: 0.50

20021207
	Convert to new build project.
	Update libraries.
	Correct load ordering for ssl programs.

20030118
	Option to set TCP environment variables.
	Option to set SSL environment variables a la mod_perl.
	Added sslperl.
	Correct closure handling in ssl_io.c.

20030128
	Correct setting of TCPREMOTEINFO.

20040127
	Improved env tracking in handlers.
	Updated tests.

20040204
	Clean up signal handling, process usage.

20040209
	Added nN options to sslclient.

20040313
	Added conf-ssl.
	Version 0.65.

20040317
	Delete conf-home.
	Permit nN opts in sslclient.
	Version 0.66.

20040320
	Allow sslclient to avoid server verification.
	Support cipher selection.
	Correct execution botch in sslperl.
	Clean up ssl_io and signal interaction.
	Version 0.67.

20040502
	Read SSL setup info before dropping privilege.
	Version 0.68.

20050417
	Don't fail on setsid if already group leader (Gabriel Russell).
	Add compatibility macros for older Perl.
	Don't leak pointer returned from eval_pv.
	Support for partial testing.
	Remove dependency on daemontools from rts.

20050515
	Partial-install bug fix in package/upgrade.

20050605
	Don't lose file descriptors in sslclient (Paul Jarc).

20050610
	Close socket for deny rule (Fred Lindberg).

20050703
	Revise ssl_io.
	Free ssl object in sslhandle.c.
	Add progtimeout option.

20050717
	Version 0.70.

20090811 
	Included ucspi-ssl-0.70_ucspitls-0.6.patch (STARTTLS support)
	originally designed and provided by Scott Gifford (FEH).

20100319
	Added Certchain support for sslserver and sslclient (FEH).

20120217
	Integration and added man-pages (FEH).
	Version 0.80.

20120921
	Synced with ucspi-tcp6-0.95.
	Version 0.82

20121003
	Fixed integration bug in ssl_very.c.
	Included patches from Peter Conrad.
	Version 0.83.

20121005
	Bug fix in sslserver. Several small
	corrections.
	Version 0.83a

20121019
	Fix for lorge X509 serial numbers on x86 (tx. Peter Conrad).
	SAN DNSname has precedence over CN in subject.
	Re-edited man pages and rts tests.
	Version 0.84.

	


src/sslclient.c
#include <unistd.h>
#include <sys/types.h>
#include <sys/param.h>
#include <netdb.h>
#include <arpa/inet.h>
#include "ucspissl.h"
#include "sig.h"
#include "exit.h"
#include "sgetopt.h"
#include "uint16.h"
#include "fmt.h"
#include "scan.h"
#include "str.h"
#include "ip.h"
#include "uint16.h"
#include "socket.h"
#include "fd.h"
#include "stralloc.h"
#include "buffer.h"
#include "getln.h"
#include "error.h"
#include "strerr.h"
#include "pathexec.h"
#include "timeoutconn.h"
#include "remoteinfo.h"
#include "auto_cafile.h"
#include "auto_cadir.h"
#include "auto_ciphers.h"
#include "byte.h"
#include "ndelay.h"
#include "wait.h"

#define FATAL "sslclient: fatal: "
#define CONNECT "sslclient: unable to connect to "

void nomem(void) {
  strerr_die2x(111,FATAL,"out of memory");
}
void env(const char *s,const char *t) {
  if (!pathexec_env(s,t)) nomem();
}
int error_warn(const char *x) {
  if (!x) return 0;
  strerr_warn2("sslclient: ",x,0);
  return 0;
}
void usage(void) {
  strerr_die1x(100,"sslclient: usage: sslclient \
[ -346hHrRdDqQveEsSnNxX ] \
[ -i localip ] \
[ -p localport ] \
[ -T timeoutconn ] \
[ -l localname ] \
[ -t timeoutinfo ] \
[ -a cafile ] \
[ -A cadir ] \
[ -c certfile ] \
[ -z ciphers ] \
[ -k keyfile ] \
[ -V verifydepth ] \
[ -w progtimeout ] \
host port program");
}

int verbosity = 1;
int flagdelay = 0;
int flagremoteinfo = 1;
int flagremotehost = 1;
int flag3 = 0;
int flagsslenv = 0;
int flagtcpenv = 0;
unsigned long itimeout = 26;
unsigned long ctimeout[2] = { 2, 58 };
unsigned int progtimeout = 3600;

const char *forcelocal = 0;

socket_address local, remote;

char *hostname;
int flagname = 1;
int flagservercert = 1;
static stralloc moreaddresses;

static stralloc tmp;
char strnum[FMT_ULONG];
static stralloc ipstr;

char seed[128];

char bspace[16];
buffer b;

SSL_CTX *ctx;
const char *certfile = 0;
const char *keyfile = 0;
const char *cafile = auto_cafile;
const char *cadir = auto_cadir;
const char *ciphers = auto_ciphers;
stralloc password = {0};
int match = 0;
int verifydepth = 1;

int pi[2];
int po[2];
int pt[2];

void read_passwd() {
  if (!password.len) {
    buffer_init(&b,buffer_unixread,3,bspace,sizeof bspace);
    if (getln(&b,&password,&match,'\0') == -1)
      strerr_die2sys(111,FATAL,"unable to read password: ");
    close(3);
    if (match) --password.len;
  }
}

int passwd_cb(char *buf,int size,int rwflag,void *userdata) {
  if (size < password.len)
    strerr_die2x(111,FATAL,"password too long");

  byte_copy(buf,password.len,password.s);
  return password.len;
}

int main(int argc,char * const *argv) {
  unsigned long u;
  int opt;
  const char *x, *portname, *localname = NULL, *portlocal = NULL;
  int j;
  int s;
  int cloop;
  SSL *ssl;
  int wstat;
  struct addrinfo *to_bind = NULL, *to_connect = NULL, hints = {0}, *bindme;
  uint16 port;

  close(6);
  close(7);
  sig_ignore(sig_pipe);
 
  hints.ai_family = AF_UNSPEC;
  while ((opt = getopt(argc,argv,"dDvqQhHrRi:p:t:T:l:a:A:c:z:k:V:3eEsSnN0xXw:46")) != opteof)
    switch(opt) {
      case 'd': flagdelay = 1; break;
      case 'D': flagdelay = 0; break;
      case 'v': verbosity = 2; break;
      case 'q': verbosity = 0; break;
      case 'Q': verbosity = 1; break;
      case 'l': forcelocal = optarg; break;
      case 'H': flagremotehost = 0; break;
      case 'h': flagremotehost = 1; break;
      case 'R': flagremoteinfo = 0; break;
      case 'r': flagremoteinfo = 1; break;
      case 't': scan_ulong(optarg,&itimeout); break;
      case 'T': j = scan_ulong(optarg,&ctimeout[0]);
		if (optarg[j] == '+') ++j;
		scan_ulong(optarg + j,&ctimeout[1]);
		break;
      case 'w': scan_uint(optarg,&progtimeout); break;
      case 'i': localname = optarg; break;
      case 'p': portlocal = optarg; break;
      case 'a': cafile = optarg; break;
      case 'A': cadir = optarg; break;
      case 'c': certfile = optarg; break;
      case 'z': ciphers = optarg; break;
      case 'k': keyfile = optarg; break;
      case 'V': scan_ulong(optarg,&u); verifydepth = u; break;
      case '3': flag3 = 1; break;
      case 'S': flagsslenv = 0; break;
      case 's': flagsslenv = 1; break;
      case 'E': flagtcpenv = 0; break;
      case 'e': flagtcpenv = 1; break;
      case 'N': flagname = 0; break;
      case 'n': flagname = 1; break;
      case 'x': flagservercert = 1; break;
      case 'X': flagservercert = 0; break;
      case '4': hints.ai_family = AF_INET; break;
      case '6': hints.ai_family = AF_INET6; break;
      default: usage();
    }
  argv += optind;

  if (!verbosity)
    buffer_2->fd = -1;

  hostname = *argv;
  if (!hostname) usage();
  if (str_equal(hostname,"")) hostname = "127.0.0.1";
  if (str_equal(hostname,"0")) hostname = "127.0.0.1";
  j = strlen(hostname);
  if (*hostname == '[' && hostname[j-1] == ']') {
    hostname[j-1] = 0;
    hostname++;
  }

  portname = *++argv;
  if (!portname) usage();

  if (flag3) read_passwd();

  if (cafile && str_equal(cafile,"")) cafile = 0;
  if (cadir && str_equal(cadir,"")) cadir= 0;
  if (ciphers && str_equal(ciphers,"")) ciphers= 0;

  if (certfile && str_equal(certfile,"")) certfile = 0;
  if (keyfile && str_equal(keyfile,"")) keyfile = 0;

  if (!*++argv) usage();

  hints.ai_socktype = SOCK_STREAM;
  if (hints.ai_family == AF_UNSPEC) {
    hints.ai_flags |= AI_ADDRCONFIG | AI_V4MAPPED;
  }
  if (localname || portlocal) {
    errno = getaddrinfo(localname, portlocal, &hints, &to_bind);
    if (errno)
      strerr_die5x(111,FATAL,"temporarily unable to figure out IP address for ",localname,": ",gai_strerror(errno));
    if (!to_bind)
      strerr_die5x(111,FATAL,"no address for ",localname?localname:"''","/",portlocal?portlocal:"0");
  }

  errno = getaddrinfo(hostname, portname, &hints, &to_connect);
  if (errno)
    strerr_die5x(111,FATAL,"temporarily unable to figure out IP address for ",hostname,": ",gai_strerror(errno));
  if (!to_connect)
    strerr_die5x(111,FATAL,"no address for ",hostname,"/",portname);
 
  if (!to_connect->ai_next) {
    ctimeout[0] += ctimeout[1];
    ctimeout[1] = 0;
  }

  s = -1;
  if (!stralloc_copys(&moreaddresses,"")) nomem();
  for (cloop = 0;cloop < 2;++cloop) {
    for (j=0, hints.ai_next = to_connect; hints.ai_next; hints.ai_next = hints.ai_next->ai_next, j++) {
      bindme = to_bind;
      while (bindme && bindme->ai_family != hints.ai_next->ai_family)
       bindme = bindme->ai_next;
      if (!bindme && to_bind) { continue; }
      if (cloop && !moreaddresses.s[j]) { continue; }
      if (hints.ai_next->ai_family > AF_MAX) { continue; }
      s = socket_tcp(hints.ai_next->ai_family, hints.ai_next->ai_protocol);
      if (s == -1)
        strerr_die2sys(111,FATAL,"unable to create socket: ");
      if (bindme && socket_bind(s,bindme) == -1)
        strerr_die2sys(111,FATAL,"unable to bind socket: ");
      byte_copy(&remote, hints.ai_next->ai_addrlen, hints.ai_next->ai_addr);
      if (timeoutconn(s,&remote,ctimeout[cloop]) == 0)
        goto CONNECTED;
      close(s);
      if (!cloop && ctimeout[1] && (errno == error_timeout)) {
	if (!stralloc_catb(&moreaddresses,"\001",1)) nomem();
      }
      else {
	if (!stralloc_catb(&moreaddresses,"",1)) nomem();
	uint16_unpack_big(remote.sa4.sin_family == AF_INET ? &(remote.sa4.sin_port)
							   : &(remote.sa6.sin6_port),
			  &port);
	strnum[fmt_ulong(strnum,port)] = 0;
	if ((opt = ip_fmt(&ipstr,(socket_address *) hints.ai_next->ai_addr)))
	  strerr_die3x(111, FATAL, "unable to print local ip: ",gai_strerror(opt));
        strerr_warn5(CONNECT,ipstr.s," port ",strnum,": ",&strerr_sys);
      }
    }
  }

  _exit(111);

  CONNECTED:

  if (socket_local(s,&local,&port) == -1)
    strerr_die2sys(111,FATAL,"unable to get local address: ");

  strnum[fmt_ulong(strnum,port)] = 0;
  env("SSLLOCALPORT",strnum);
  if (flagtcpenv) env("TCPLOCALPORT",strnum);
  if (ip_fmt(&ipstr,&local)) nomem();
  env("SSLLOCALIP",ipstr.s);
  if (flagtcpenv) env("TCPLOCALIP",ipstr.s);

  x = forcelocal;
  if (!x)
    if (dns_name(&tmp,&local) == 0) {
      x = tmp.s;
    }
  env("SSLLOCALHOST",x);
  if (flagtcpenv) env("TCPLOCALHOST",x);

  if (socket_remote(s,&remote,&port) == -1)
    strerr_die2sys(111,FATAL,"unable to get remote address: ");

  strnum[fmt_ulong(strnum,port)] = 0;
  env("SSLREMOTEPORT",strnum);
  if (flagtcpenv) env("TCPREMOTEPORT",strnum);
  if (ip_fmt(&ipstr,&remote)) nomem();
  env("SSLREMOTEIP",ipstr.s);
  /* If ipstr.s contain ':' colon character will assume it is IPv6 */
  if (byte_chr(ipstr.s, ipstr.len, ':') < ipstr.len)
    env("PROTO","SSL6");
  else
    env("PROTO","SSL");
  if (flagtcpenv) env("TCPREMOTEIP",ipstr.s);
  if (verbosity >= 2)
    strerr_warn4("sslclient: connected to ",ipstr.s," port ",strnum,0);

  x = 0;
  if (flagremotehost)
    if (dns_name(&tmp,&remote) == 0) {
      x = tmp.s;
    }
  env("SSLREMOTEHOST",x);
  if (flagtcpenv) env("TCPREMOTEHOST",x);

  x = 0;
  if (flagremoteinfo)
    if (remoteinfo(&tmp,&remote,&local,itimeout) == 0) {
      if (!stralloc_0(&tmp)) nomem();
      x = tmp.s;
    }
  env("SSLREMOTEINFO",x);
  if (flagtcpenv) env("TCPREMOTEINFO",x);

  ctx = ssl_client();
  ssl_errstr();
  if (!ctx)
    strerr_die2x(111,FATAL,"unable to create SSL context");

  switch (ssl_certkey(ctx,certfile,keyfile,passwd_cb)) {
      case -1: strerr_die2x(111,FATAL,"unable to load certificate");
      case -2: strerr_die2x(111,FATAL,"unable to load key pair");
      case -3: strerr_die2x(111,FATAL,"key does not match certificate");
      default: break;
  }
  
  if (!ssl_ca(ctx,cafile,cadir,verifydepth))
    strerr_die2x(111,FATAL,"unable to load CA list");

  if (!ssl_ciphers(ctx,ciphers))
    strerr_die2x(111,FATAL,"unable to set cipher list");

  ssl = ssl_new(ctx,s);
  if (!ssl) strerr_die2x(111,FATAL,"unable to create SSL instance");

  for (cloop = 0;cloop < 2;++cloop) {
    if (!ssl_timeoutconn(ssl,ctimeout[cloop])) goto SSLCONNECTED;
    if (!cloop && ctimeout[1]) continue;
    strerr_warn2(FATAL,"unable to SSL connect:",&strerr_sys);
    ssl_error(error_warn);
  }

  _exit(111);

  SSLCONNECTED:

  ndelay_off(s);

  if (verbosity >= 2)
    strerr_warn1("sslclient: ssl connect",0);

  if (flagservercert)
    switch(ssl_verify(ssl,flagname ? hostname : 0)) {
      case -1:
	strerr_die2x(111,FATAL,"unable to verify server certificate");
      case -2:
	strerr_die2x(111,FATAL,"no server certificate");
      case -3:
	strerr_die3x(111,FATAL,"certificate name does not match server hostname: ",hostname);
      default: break;
    }

  if (!flagdelay)
    socket_tcpnodelay(s); /* if it fails, bummer */

  if (pipe(pi) == -1) strerr_die2sys(111,FATAL,"unable to create pipe: ");
  if (pipe(po) == -1) strerr_die2sys(111,FATAL,"unable to create pipe: ");
  if (pi[0] == 7) {
    if (pipe(pt) == -1) strerr_die2sys(111,FATAL,"unable to create pipe: ");
    close(pi[0]); close(pi[1]);
    pi[0] = pt[0]; pi[1] = pt[1];
  }
  if (po[1] == 6) {
    if (pipe(pt) == -1) strerr_die2sys(111,FATAL,"unable to create pipe: ");
    close(po[0]); close(po[1]);
    po[0] = pt[0]; po[1] = pt[1];
  }

  switch(opt = fork()) {
    case -1:
      strerr_die2sys(111,FATAL,"unable to fork: ");
    case 0:
      break;
    default:
      close(pi[0]); close(po[1]);
      if (ssl_io(ssl,pi[1],po[0],progtimeout)) {
	strerr_warn2(FATAL,"unable to speak SSL:",&strerr_sys);
	ssl_error(error_warn);
	ssl_close(ssl);
	wait_pid(&wstat,opt);
	_exit(111);
      }
      ssl_close(ssl);
      if (wait_pid(&wstat,opt) > 0)
	_exit(wait_exitcode(wstat));
      _exit(0);
  }
  ssl_close(ssl); close(pi[1]); close(po[0]);

  if (flagsslenv && !ssl_client_env(ssl,0)) nomem();

  if (fd_move(6,pi[0]) == -1)
    strerr_die2sys(111,FATAL,"unable to set up descriptor 6: ");
  if (fd_move(7,po[1]) == -1)
    strerr_die2sys(111,FATAL,"unable to set up descriptor 7: ");
  sig_uncatch(sig_pipe);

  pathexec(argv);
  strerr_die4sys(111,FATAL,"unable to run ",*argv,": ");
  return 0; /* never happens, but avoids compile warning */
}
